File: authed.py
import requests
import jwt

class Authed:
    baseUri = "https://api.authed.io/app"

    def __init__(self, _id, access, userKey):
        self.appId = _id
        self.access = access
        self.userKey = userKey
        self.userSession = None
        self.appSession = None

    def _call_api(self, endpoint, body, _headers):
        req = requests.post(endpoint, data=body, headers=_headers)
        return req.json()

    def verify(self):
        json = self._call_api(baseUri + "/verify/" + self.appId, self.access, None)
        if json['session'] != None:
            self.appSession = json['session']
            return True
        else:
            raise Exception('{}: {}'.format(json['status'], json['message']))

    def login(self, email, password):
        headers = { 'session': self.appSession }
        body = {
            'email': email,
            'password': password,
        }
        json = self._call_api(baseUri + "/login", body, headers)

        if json['status'] == 200:
            self.userSession = json['userSesion']
            return True
        else:
            raise Exception('{}: {}'.format(json['status'], json['message']))

    
    def register(self, email, password, _license):
        headers = { 'session': self.appSession }
        body = {
            'email': email,
            'password': password,
            'licenseCode': _license
        }

        json = self._call_api(baseUri + "/register", body, headers)

        if json['status'] == 200:
            return True
        else:
            raise Exception('{}: {}'.format(json['status'], json['message']))

    def renew(self, email, _license):
        headers = { 'session': self.appSession }
        body = {
            'email': email,
            'licenseCode': _license
        }

        json = self._call_api(baseUri + "/renew", body, headers)

        if json['stats'] == 200:
            return True
        else: 
            raise Exception('{}: {}'.format(json['status'], json['message']))

    def generate_license(self, secret, prefix, amount, level, time):
        headers = { 'session': self.appSession }
        body = {
            'secret': secret,
            'prefix': prefix,
            'amount': amount,
            'level': level,
            'time': time,
        }

        json = self._call_api(self, body, headers)

        if json['status'] == 200:
            return json['message']
        else:
            raise Exception('{}: {}'.format(json['status'], json['message']))

    def update_user(self, field, value):
        headers = { 'session': self.appSession }
        body = {
            'userSession': self.userSession,
            'field': field,
            'value': value
        }

        json = self._call_api(baseUri + "/users/set", body, headers)

        if json['status'] == 200:
            return True
        else:
            raise Exception('{}: {}'.format(json['status'], json['message']))

    def update_app(self, field, value, secret):
        headers = { 'session': self.appSession }
        body = {
            'secret': secret,
            'value': value
        }

        json = self._call_api(baseUri + "/set/" + field, body, headers)

        if json['status'] == 200:
            return True
        else:
            raise Exception('{}: {}'.format(json['status'], json['message']))

    # worning this is only going to work if you set a custom user key on Authed
    def verifyUserSession(self):
        return jwt.decode(self.userSession, self.userKey, algorithms=['HS256'])

File: Authed.cs
using System;
using System.Collections.Generic;
using System.Net;
using Newtonsoft.Json;
using System.Timers;

class Authed {
    public internal string AppSession;
    public internal string UserSession;
    public internal string AppID;
    public internal string UserKey;
    public internal string Access;
    private internal string baseUri = "https://api.authed.io/app";

    public Authed(string id, string access, string userKey) {
        this.AppID = id;
        this.UserKey = userKey;
        this.Access = access;
    }

    private dynamic CallApi(string url, Dictionary<string, string> body, Dictionary<string, string> headers) {
        WebClient wc = new WebClient();
        wc.Headers[HttpRequestHeader.ContentType] = "application/x-www-form-urlencoded";
        string bodyString = "";
        foreach(KeyValuePair<string, string> entry in body)
        {
            bodyString.Length > 0 ? bodyString += $"{entry.Key}={entry.Value}" : bodyString = $"{entry.Key}={entry.Value}";
        }
        foreach(KeyValuePair<string, string> entry in headers) {
            wc.Headers.Add(entry.Key, entry.Value);
        }
        return new JsonConvert.DeserializeObject(wc.UploadString(url, bodyString));
    }

    private void DoAppSessionCheck() {
        if(this.AppSession == null) throw new System.Exception("Please verify before logging in!");
    }
    
    private void DoUserSessionCheck() {
        if(this.UserSession == null) throw new System.Exception("Please login before accessing this function!");
    }

    private void StartUserSesionTimer() {
        System.Timers.Timer sessTimer = new System.Timers.Timer(25000);
        sessTimer.Elapsed += OnTimedEvent;
        sessTimer.Enabled = true;
        sessTimer.Start();
    }

    public boolean VerifyUserSession() {
        var headers = new Dictionary<string, string>()
        {
            { "session", this.AppSession }
        };

        var body = new Dictionary<string, string>()
        {
            { "userSession", this.UserSession }
        };

        dynamic json = this.CallApi($"{this.baseUri}/session", body, headers);
        if(json.userSession != null) {
            // Use Jose.JWT to verify this if you want. Its more "secure"
            this.UserSession = json.userSession;
            // CHANGE
        } else {
            // Bad exit, invalid user
            Environment.Exit(99);
            // Do something here
            // CHANGE
        }
    }

    private static void OnTimedEvent(Object source, ElapsedEventArgs e)
    {
        this.VerifyUserSession();
    }

    public boolean Verify() {
        var body = new Dictionary<string, string>() 
        {
            { "access",  this.Access }
        };

        dynamic json = this.CallApi($"{this.baseUri}/verify/{this.AppID}", body, null);
        if(json.session != null) {
            this.AppSession = json.session;
            return true;
        } else {
            throw new System.Exception(json.message);
        }
    }
    
    public boolean Login(string email, string password) {
        this.DoAppSessionCheck();

        var headers = new Dictionary<string, string>()
        {
            { "session", this.AppSession },
        };

        var body = new Dictionary<string, string>()
        {
            { "email", email },
            { "password", password}
        };

        dynamic json = this.CallApi($"{this.baseUri}/login", body, headers);

        if(json.userSession != null) {
            this.UserSession = json.userSession;
            this.StartUserSesionTimer();
            return true;
        } else {
            throw new System.Exception(json.message);
        }
    }

    public boolean Register(string email, string password, string license) {
        this.DoAppSessionCheck();

        var headers = new Dictionary<string, string>()
        {
            { "session", this.AppSession },
        };

        var body = new Dictionary<string, string> ()
        {
            { "email", email },
            { "password", password },
            { "licenseCode", license }
        };

        dynamic json = this.CallApi($"{this.baseUri}", body, headers);

        if(json.status == 200) return true;
        else throw new System.Exception(json.message);
        
    }

    public boolean Renew(string email, string license) {
        this.DoAppSessionCheck();
        var headers = new Dictionary<string, string>()
        {
            { "session", this.AppSession }
        };

        var body = new Dictionary<string, string>()
        {
            { "email", email },
            { "licenseCode", license }
        };

        dynamic json = this.CallApi($"{this.baseUri}/renew", body, headers);

        if(json.status == 200) return true;
        else throw new System.Exception(json.message);
    }

    public boolean UpdateUser(string field, string value) {
        this.DoAppSessionCheck();
        this.DoUserSessionCheck();

        var headers = new Dictionary<string, string>()
        {
            { "session", this.AppSession }
        };

        var body = new Dictionary<string, string>()
        {
            { "userSession", this.UserSession },
            { "field", field },
            { "value", valie },
        };

        dynamic json = this.CallApi($"{this.baseUri}/users/set", body, headers);

        if(json.status == 200) return true;
        else throw new System.Exception(json.message);
    }

    public boolean UpdateApp(string secret, string field, string value) {
        this.DoAppSessionCheck();

        var headers = new Dictionary<string, string>()
        {
            { "session", this.AppSession }
        };

        var body = new Dictionary<string, string>()
        {
            { "secret", secret },
            { "value", value }
        };

        dynamic json = this.CallApi($"{this.baseUri}/set/${field}", body, headers);

        if(json.status == 200) return true;
        else throw new System.Exception(json.message);
    }

    public GenerateLicense(string secret, string prefix, int level, int time, int amount) {
        this.DoAppSessionCheck();

        var headers = new Dictionary<string, string>()
        {
            { "session", this.AppSession }
        };

        var body = new Dictionary<string, string>()
        {
            { "secret", secret },
            { "prefix", prefix },
            { "level", level },
            { "time", time },
            { "amount", amount }
        };

        dynamic json = this.CallApi($"{this.baseUri}/generate/license", body, headers);

        if(json.status == 200) return json.message;
        else throw new System.Exception(json.message);
    }
}